compass_app/app/lib/utils/view_command.dart

```diff
@@ -0,0 +1,112 @@
+typedef TypedCommand<T> = void Function(T);
+
+/// Interface to a type of commands that take an argument of type [T] and
+/// returns a value of type [R].
+typedef TypedCommandWithReturn<R, T> = R Function(T);
+
+/// Interface to a type of commands that take no arguments and executes
+/// asynchronously.
+///
+/// [ViewCommand] or [TypedCommand] can still call asynchronous commands. But
+/// certain Flutter widgets expect a [Future]. For example, a [RefreshIndicator]
+/// takes a [RefreshCallback]. And an [AsyncCommand] can be used.
+typedef AsyncCommand = Future<void> Function();
+
+/// Interface to a type of commands that takes a [BuildContext] and executes
+/// asynchronously.
+typedef AsyncContextCommand = Future<void> Function(BuildContext);
+
+/// Interface to a type of commands that take an argument of type [T] and
+/// executes asynchronously.
+///
+/// See [ContextCommand] for hints on importing Flutter symbols, such as
+/// [BuildContext].
+typedef AsyncTypedCommand<T> = Future<void> Function(T);
+
+/// Interface to a type of commands that take no arguments, execute
+/// asynchronously and returns a value of type [R].
+typedef AsyncCommandWithReturn<R> = Future<R> Function();
+
+/// Interface to a type of commands that take an argument of type [T], execute
+/// asynchronously and returns a value of type [R].
+typedef AsyncTypedCommandWithReturn<R, T> = Future<R> Function(T);
+
+/// Interface to a type of commands that take a [BuildContext].
+///
+/// Quite common in Flutter widget code, this is just an alias to
+/// [TypedCommand<BuildContext>].
+///
+/// Example: a Flutter navigation operation.
+///
+/// Note: when referenceing [BuildContext] in a built value, there can be a
+/// compiler error about an amgibuous `Builder` symbol. This is because the
+/// Flutter framework and package:built_value both exports a `Builder` symbol.
+/// To work around this problem, `hide` the `Builder` symbol when importing
+/// Flutter libraries in your built value file. For example:
+///
+/// ```dart
+/// import 'package:built_value/built_value.dart';
+/// import 'package:flutter/widgets.dart' hide Builder;
+///
+/// part 'view_model.g.dart';
+///
+/// abstract class ViewModel implements Built<...> {
+///   ...
+///   /// Or simply [ContextCommand].
+///   TypedCommand<BuildContext> get myCommand;
+/// }
+/// ```
+///
+/// With this approach, the [Builder] widget from Flutter is unavailable in this
+/// file. In the extremely rare case that your built value implementation needs
+/// access to Flutter's [Builder], you may choose to provide an import prefix
+/// when importing Flutter:
+///
+/// ```dart
+/// import 'package:built_value/built_value.dart';
+/// import 'package:flutter/widgets.dart' as flutter;
+///
+/// part 'view_model.g.dart';
+///
+/// abstract class ViewModel implements Built<...> {
+///   ...
+///   TypedCommand<flutter.BuildContext> get myCommand;
+/// }
+/// ```
+typedef ContextCommand = void Function(BuildContext);
```

Comment:
This is the most interesting part. Presumably this is useful for moving routing logic outside of the widgets, and for getting device information if necessary. 

Looking at this example, my updates in this PR differ from Stadias in that the methods are being directly accessed from the ViewModel. Their "ViewModels" don't contain logic if I understand correctly, but they have a library that binds ViewModels to Components, so I don't know that we'd want to concern ourselves with that.

Reply:
One thing I find interesting with this, is that the `BuildContext` is taken outside Widgets and into the app logic. For example, you could start to have navigation code or showing snackbars implemented in the ViewModel side... I think this is very dangerous if used incorrectly and we should not encourage this. Instead, code that uses BuildContext should remain in the Widget.